include/clap/ext/draft/scratch-memory.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include "../../plugin.h"
+
+// This extension lets the plugin request "scratch" memory
+// from the host. Scratch memory can be accessed during the
+// `process()` callback, but is not persistent between callbacks.
+//
+// The motivation for this extension is to allow the plugin host
+// to "share" a single scratch buffer across multiple plugin
+// instances.
+//
+// For example, imagine the host needs to process three plugins
+// in sequence, and each plugin requires 10K of scratch memory.
+// If each plugin pre-allocates its own scratch memory, then 30K
+// of memory is being allocated in total. However, if each plugin
+// requests 10K of scratch memory from the host, then the host can
+// allocate a single 10K scratch buffer, and make it available to all
+// three plugins.
+//
+// On memory-constrained platforms, this optimization may allow for
+// more plugins to be used simultaneously. On platforms with lots
+// of memory, this optimization may improve CPU cache usage.
+
+static CLAP_CONSTEXPR const char CLAP_EXT_SCRATCH_MEMORY[] = "clap.scratch-memory/1";
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct clap_host_scratch_memory {
+   // Asks the host for certain amount of scratch memory.
+   // If the host is unable to provide the memory, it should
+   // return "false".
+   //
+   // Note that any memory the host allocates to satisfy
+   // the requested scratch size can be de-allocated by the
+   // host when the plugin is de-activated.
+   //
+   // [main-thread & being-activated]
+   bool(CLAP_ABI *pre_reserve)(const clap_host_t *host, size_t scratch_size_bytes);
```

Comment:
Do we ever use `size_t` in any other extension? -> `uint?_t` ?

Reply:
Good point, `uint32_t` would do the job I think.

Reply:
Why not use uint64_t which is size_t in most of our systems? The host has the option to return no for values out of bounds

Reply:
I don't have a preference for `uint32_t` vs `uint64_t`... but Trinitou is right that CLAP doesn't use `size_t` anywhere else, so I don't think we should use it here.

Reply:
A scratch size bigger than 4 GB would be problematic I suppose.
Anyway regardless of the type, many host will likely have their own threshold.